.gitattributes

```diff
@@ -13,4 +13,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# Do not include KEYS in archived source releases
 /KEYS export-ignore 
+
+# Do not convert to CRLF on Windows since tests require LF files
+/**/runtime2/examples/** text=auto eol=lf
```

Comment:
It might also be worth punting this whitespace and checking failing test if generated not updated to another PR, this feels like it could become a big change, and it would be good to get this corner case changes in.

Reply:
Steve, I just saw your comments.  Yes, your sbt code works around the version issue, but so does running the unit test with "sbt test" without needing any changes to build.sbt.  I think having a unit test is nice because it offers a way to quickly test and debug changes to the code generator using a couple of larger example schemas.

And yes, your generator changes look like they would allow the test to pass regardless of which line endings the git checkout uses.  However, there are a lot of multiline strings in runtime2 and we would need to make these `stripMargin.replace("\r\n", "\n")` changes to every place. I think it's a lot easier and simpler to add one line in .gitattributes telling git to check out all of daffodil-runtime as LF regardless of platform or git config settings.  However, I need a second+1 anyway, so Mike (@mbeckerle), will you please tell us which way you prefer?

I also wonder which would give us fewer problems in the long run, telling git to check out all of daffodil as LF with `* text=auto eol=LF` or continuing to call `if (windows)`, `replace("\r\n", "\n")`, `replace("\n", System.lineSeparator)`, `mkString(System.lineSeparator)`, and so on.  I suspect it's probably not worth making major changes to line separators in files or code and we should keep changing as few files and lines of code as possible.  The tests have all passed now, so I think checking out daffodil-runtime2 as LF without changing any other part of daffodil has the least risk and will spare daffodil-runtime2 from needing explicit code to manage line separators in the future.

Reply:
Another thought on the test approach, if the example files are out of date, the first time you run tests you'll get a failure, but the second time the test will pass, even though the user hasn't changed anything. That feels like potentially confusing behavior. And I think that's because a test is updated the expected results. It feels wrong tests to be updating expected results.

I'm fine doing something other than the `genExamples` approach if you think it's too slow--daffodil-cli/stage does take some time, especially when making updates. But I'm not sure tests are the right place for it.


Whitespace issue: If we choose to update .gitattributes to have daffodil-runtime2/ to disable autocrlf, then I think we should just do it for everything. Having some source that is always LF and some that is OS dependent feels wrong. That said, as dumb as I think Git's decision is to enable autocrlf, I also don't want to make development harder for people on windows. Maybe there are some tools that really do need this autocrlf feature, so I'm not sure disabling it for .scala files is safe? I'm not really sure...

One last thought regarding whitespace, we could minimize code changes by adding some implicit classes:
```scala
implicit class StringNewLine(s: String) {
  def stripMarginNL: String = s.stripMargin.replace("\r\n", "\n").replace("\n", System.lineSeparator)
}

implicit class SeqNewLine[T](s: Seq[T]) {
  def mkStringNL: String = s.mkString(System.lineSeparator)
}
```

Then multiline strings become something like
```scala
val foo =
  """|multi
     |line
     |string""".stripMarginNL
```
and Seqs become

```scala
val bar = Seq("multi", "line", "string").mkStringNL
```
There would need to be a big change to switch it over, but the new code would look fairly clean and would ensure consistent newlines.

Reply:
OK, we are discussing two separate issues now:

1. Should test_updateGeneratedCodeExamples fail every time or only the first time?
2. Should the C code generator write a) only LFs or b) System.lineSeparator into generated code files regardless of which line endings are used in source files?

For 1), I don't think we can separate the test and the update without adding an extra step (the update command) to the code contributor's workflow.  Instead of simply running "sbt test" until no test fails, the contributor will have to run "sbt genExamples" and then "sbt test" again to verify the tests pass now.  Do we really want to add more friction to the code contributor's workflow?  

I say the convenience of combining the test and the update outweighs the strangeness of the test failing only the first time.  The test may seem like it will fail only once, but it will invariably fail every time when run on a fresh checkout of the repository on a GitHub Actions runner.  The test in CI will prevent anyone from committing a change to the code generator without running "sbt test" first.  

At the same time, I do think we need to expand the assert's message so that any code contributor will understand what is going on, especially if they're merely publishing a new daffodil release (even changing daffodil's version in build.sbt without running "sbt test" will make the CI job fail).  "Examples have been updated!" is too terse for GitHub Actions; we probably should write something like "You need to update the examples!  Run `sbt test` and commit the updated example code files."  Suggestions for better phrasing would be welcome.

For 2), I could make an even smaller change if we want the C code generator to write only LFs or System.lineSeparator regardless of source files' line endings.  I would modify CodeGeneratorState's generatedCodeHeader and generatedCodeFile functions to replace CRLF and LF before returning the final strings to be written out to files.  That's only two new lines of code, but we would have to replace LF with System.lineSeparator before we could remove the new line from .gitattributes.  Unfortunately, if we remove the new line from .gitattributes, the test still could fail in a code contributor's checkout due to the contributor's git config settings.  The test will work in GitHub Actions, but it won't work for all code contributors running it because they could have checked out the example code files on Windows with either LF or CRLF line endings depending on their settings.  The best thing to do would be to leave the new line but change it to  `* text=auto`, which causes git to check out all text-like files with System.lineSeparator line endings regardless of git config settings.  Then the examples and the generated files will match each other - they'll always be System.lineSeparator on GitHub Actions and in anyone's checkout.  Files that are already CRLF in git will remain CRLF on checkout with this new line as well.

John (@jw3) or anyone else, please weigh in too (I've been waiting a long time for a second +1).

Reply:
I've implemented the changes I talked about above ("* text=auto" and System.lineSeparator) and the tests still pass, so this is the final version I think should be merged.  Still need a second +1.